Add Custom Schedule #13

growdev opened the pull request:
For #13 

Tasks: 

- [x] Add a meta box to product set CPT
- [x] Save the schedule to custom meta `_dfrps_update_schedule`
- [x] Make updates to `dfrps_get_next_update_time()`
- [ ] Testing

EricBusch commented on classes/class-dfrps-cpt.php:
```diff
@@ -296,6 +296,16 @@ function add_custom_meta_boxes() {
 
+		add_meta_box(
+			'dfrps_cpt_schedule_metabox',
+			_x( 'Custom Update Schedule', DFRPS_DOMAIN ),
```
@growdev  Unfortunately I made the mistake of using the constant `DFRPS_DOMAIN` for the `$domain` argument in translation functions. Unfortunately, I learned later, that this method actually doesn't work. So we should use `'datafeedr-product-sets'` instead of `DFRPS_DOMAIN`. I thought I had fixed that in this plugin but I guess I only fixed this issue the API plugin.  But, it's not a huge deal because we don't ship any translation files anyway. ☹️  Thanks!

growdev replied:
Ah, good point. I noticed phpcs added this as a warning. I'll add a new issue for this so this PR doesn't get messy.

EricBusch commented on classes/class-dfrps-cpt.php:
```diff
@@ -349,6 +359,141 @@ function cpt_dashboard_metabox( $post, $box ) {
+
+	    for ( $i=1; $i < 29; $i++ ) {
+            $selected = 'Day of the month' === $settings['interval'] && in_array( $i, $settings['days'] ) ? 'selected' : '';
+            switch ( $i ){
```
Instead of a hardcoded switch statement here we could rely on `date_i18n()` which has the added benefit of translating the suffix into the user's local language.

```php
for ( $i = 1; $i < 29; $i ++ ) {
    $selected = 'Day of the month' === $settings['interval'] && in_array( $i, $settings['days'] ) ? 'selected' : '';
    $date     = sprintf( 'January %d, 2000', $i );
    
    $html .= '<option value="' . $i . '" ' . $selected . '>Every ' . esc_html( date_i18n( 'jS', strtotime( $date ) ) ) . '</option>';
}
```

growdev replied:
Added, thanks!

growdev commented:
## Implementation notes

- If Enabled is unchecked the meta is deleted. Reloading the page will show the meta box with "Enabled" unchecked.

- Because the multiselect select input has different options for "Day of week" vs "Day of the month" I create two inputs and show/hide based on the Interval choice. Let me know if you think this is wonky.
 
- The fields (Interval, Days, Time) have a red asterisk based on the design, but there is no validation that occurs when the CPT is saved. If a field is not filled out a default value is used: Day of week, Every Monday or 1st of month, 00:00.

## Questions
- While testing I observed some strangeness where saves don’t stick. I will make a change, see it in the database, but not see it on a page reload. The second page reload makes the change appear. I’m not sure if this is due to Firefox doing some local storage, heartbeat, auto-save, or something else. Let me know if you are able to recreate the same.

- When a Product Set is saved should the dfrps_get_next_update_time be updated using the custom schedule? 

- In helper.php I’m using a null coalescing operator, but this requires PHP 7.0. We chatted about PHP version, but let me know if I should remove this or if the null check is not needed in dfrps_get_custom_update_time()

- Let me know if there is a better algorithm for finding next schedule. What I put in  [dfrps_get_custom_update_time()](https://github.com/datafeedr/datafeedr-product-sets/pull/14/files#diff-bd1ab0a17cfbefe7c698fc25b849c09743aa461fd8d16ee90af75772150f2336R53-R125) works with my test cases, but there might be a more elegant way of doing this.

EricBusch commented:
> If Enabled is unchecked the meta is deleted. Reloading the page will show the meta box with "Enabled" unchecked.

Perfect, no need the postmeta table containing any more data than necessary. It's already abused by so many plugins.

> Because the multiselect select input has different options for "Day of week" vs "Day of the month" I create two inputs and show/hide based on the Interval choice. Let me know if you think this is wonky.

I think this is fine. It's working well.

> The fields (Interval, Days, Time) have a red asterisk based on the design, but there is no validation that occurs when the CPT is saved. If a field is not filled out a default value is used: Day of week, Every Monday or 1st of month, 00:00.

Since selecting an "interval" results in an `<option>` from the select menu already being chosen, we can remove the red asterisks because it's impossible for an option not to be chosen.

> While testing I observed some strangeness where saves don’t stick. I will make a change, see it in the database, but not see it on a page reload. The second page reload makes the change appear. I’m not sure if this is due to Firefox doing some local storage, heartbeat, auto-save, or something else. Let me know if you are able to recreate the same.

I don't notice this but I am noticing the "Next Update" column is not being updated here: WordPress Admin Area > Product Sets

> When a Product Set is saved should the dfrps_get_next_update_time be updated using the custom schedule?

Yes, because (as mentioned above) the "Next Update" column doesn't actually reflect the true update time.

> In helper.php I’m using a null coalescing operator, but this requires PHP 7.0. We chatted about PHP version, but let me know if I should remove this or if the null check is not needed in dfrps_get_custom_update_time()

Totally fine. The plugins all have `Requires PHP: 7.4` in their header fields so using null coalescing is fine (and preferred).

> Let me know if there is a better algorithm for finding next schedule. What I put in [dfrps_get_custom_update_time()](https://github.com/datafeedr/datafeedr-product-sets/pull/14/files#diff-bd1ab0a17cfbefe7c698fc25b849c09743aa461fd8d16ee90af75772150f2336R53-R125) works with my test cases, but there might be a more elegant way of doing this.

I will have a play around with this now.

EricBusch commented:
I made a few small changes. The most major thing was added `update_post_meta( $post_id, '_dfrps_cpt_next_update_time', dfrps_get_custom_update_time( $schedule ) );` to save_post()

Everything is working great! Nice work.

Can you just make a few small changes?

Can you change the array keys from "natural language" keys to obvious array keys?

For example:

`$selected = 'Day of week' === $settings['interval'] && in_array( '1', $settings['days'] ) ? 'selected' : '';`

It makes it hard to do a search and replace in the future if we just wanted to change the "Day of week" text which appears to the user.

So can you make that `day_of_week` and change `Day of the month` to `day_of_month`?

Also, let's keep the phrases consistent. So either:

**Day of week, Day of month** or **Day of the week, Day of the month**

I had a go at the algorithm but after getting into it I probably wouldn't have done it much differently.

I think the last thing I would change is make sure you use `date_i18n()` instead of `date()` when the date being used is relative to the user's WP timezone settings.

Thanks!

growdev commented:
In most recent commit I fixed the following:

- Updated array keys.
- Normalized on `Day of week`, `Day of month`.
- Removed the required tag and associated CSS.
- Changed to using `date_i18n()`
- Tested and verified Next Update is correct for both day of month and day of week calculations.

EricBusch commented:
@growdev I [added a few changes yesterday](https://github.com/datafeedr/datafeedr-product-sets/pull/14/commits/8332d99bba1562a91e6a46bf488863933b65e094). Do I need to merge that pull request before you get access to those changes?
